Improve error message when a helper/modifier/component is missing from the loose registry

The current error is:

```
app/components/workflow-tracker/index.hbs:3:17 - error TS7053: Element implicitly has an 'any' type because expression of type '"toggle"' can't be used to index type 'Globals'.
  Property 'toggle' does not exist on type 'Globals'.

3   {{on 'click' (toggle 'showing' this)}}
                  ~~~~~~
```

@chriskrycho tells me "that error actually means it’s not in your registry. We should note it as an error to rewrite."

--- a/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
+++ b/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
@@ -635,4 +635,95 @@ describe('Language Server: Diagnostic Augmentation', () => {
       ]
     `);
   });
+
+  test('failed `component` name lookup', () => {
+    project.setGlintConfig({ environment: ['ember-loose'] });
+    project.write({
+      'index.ts': stripIndent`
+        import Component from '@glimmer/component';
+
+        export default class MyComponent extends Component {
+          componentName = 'bar' as const';
+        }
+      `,
+      'index.hbs': stripIndent`
+        {{#let 'baz' as |baz|}}
+          {{#let
+            (component 'foo') 
+            (component this.componentName)
+            (component baz)
+            as |Foo Bar|
+          }}
+            {{! @glint-ignore: we don't care about errors here}}
+            <Foo /><Bar /><Baz />
+          {{/let}}
+        {{/let}}
+      `,
+    });
+
+    let server = project.startLanguageServer();
+    let diagnostics = server.getDiagnostics(project.fileURI('index.hbs'));
+
+    expect(diagnostics).toMatchInlineSnapshot(`
+      [
+        {
+          "message": "Unknown component name 'foo'. If this isn't a typo, you may be missing a registry entry for this name; see the Template Registry page in the Glint documentation for more details.
+        No overload matches this call.
+          The last overload gave the following error.
+            Argument of type 'string' is not assignable to parameter of type 'abstract new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>'.",
+          "range": {
+            "end": {
+              "character": 20,
+              "line": 2,
+            },
+            "start": {
+              "character": 15,
+              "line": 2,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2769)",
+          "tags": [],
+        },
+        {
+          "message": "The type of this expression doesn't appear to be a valid value to pass the {{component}} helper. If possible, you may need to give the expression a narrower type, for example \`'component-a' | 'component-b'\` rather than \`string\`.
+        No overload matches this call.
+          The last overload gave the following error.
+            Argument of type 'string' is not assignable to parameter of type 'abstract new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>'.",
+          "range": {
+            "end": {
+              "character": 33,
+              "line": 3,
+            },
+            "start": {
+              "character": 15,
+              "line": 3,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2769)",
+          "tags": [],
+        },
+        {
+          "message": "The type of this expression doesn't appear to be a valid value to pass the {{component}} helper. If possible, you may need to give the expression a narrower type, for example \`'component-a' | 'component-b'\` rather than \`string\`.
+        No overload matches this call.
+          The last overload gave the following error.
+            Argument of type 'string' is not assignable to parameter of type 'abstract new (...args: unknown[]) => Invokable<(args: unknown) => AcceptsBlocks<any, any>>'.",
+          "range": {
+            "end": {
+              "character": 18,
+              "line": 4,
+            },
+            "start": {
+              "character": 15,
+              "line": 4,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2769)",
+          "tags": [],
+        },
+      ]
+    `);
+  });
 });

--- a/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
+++ b/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
@@ -726,4 +726,116 @@ describe('Language Server: Diagnostic Augmentation', () => {
       ]
     `);
   });
+
+  test('direct invocation of `{{component}}`', () => {
+    project.setGlintConfig({ environment: ['ember-loose'] });
+    project.write({
+      'index.ts': stripIndent`
+        import Component from '@glimmer/component';
+
+        export interface MyComponentSignature {
+          Args: {
+            message?: string;
+          };
+          Blocks: {
+            default: [];
+          };
+        }
+
+        export default class MyComponent extends Component<MyComponentSignature> {}
+
+        declare module '@glint/environment-ember-loose/registry' {
+          export default interface Registry {
+            'my-component': typeof MyComponent;
+          }
+        }
+      `,
+      'index.hbs': stripIndent`
+        {{! inline invocation }}
+        {{component 'my-component'}}
+        {{component 'my-component' message="hi"}}
+
+        {{! block invocation }}
+        {{#component 'my-component'}}{{/component}}
+        {{#component 'my-component' message="hi"}}{{/component}}
+      `,
+    });
+
+    let server = project.startLanguageServer();
+    let diagnostics = server.getDiagnostics(project.fileURI('index.hbs'));
+
+    expect(diagnostics).toMatchInlineSnapshot(`
+      [
+        {
+          "message": "The {{component}} helper can't be used to directly invoke a component under Glint. Consider first binding the result to a variable, e.g. '{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as '<ComponentName @arg={{value}} />'.
+        Argument of type 'typeof MyComponent' is not assignable to parameter of type 'ContentValue'.",
+          "range": {
+            "end": {
+              "character": 28,
+              "line": 1,
+            },
+            "start": {
+              "character": 0,
+              "line": 1,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2345)",
+          "tags": [],
+        },
+        {
+          "message": "The {{component}} helper can't be used to directly invoke a component under Glint. Consider first binding the result to a variable, e.g. '{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as '<ComponentName @arg={{value}} />'.
+        Argument of type 'abstract new () => PartiallyAppliedComponent<{ message?: string | undefined; }, { message: string; }, AcceptsBlocks<FlattenBlockParams<{ default: { Params: { Positional: []; }; }; }>, null>>' is not assignable to parameter of type 'ContentValue'.",
+          "range": {
+            "end": {
+              "character": 41,
+              "line": 2,
+            },
+            "start": {
+              "character": 0,
+              "line": 2,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2345)",
+          "tags": [],
+        },
+        {
+          "message": "The {{component}} helper can't be used to directly invoke a component under Glint. Consider first binding the result to a variable, e.g. '{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as '<ComponentName @arg={{value}}>...</ComponentName>'.
+        Argument of type 'typeof MyComponent' is not assignable to parameter of type 'AcceptsBlocks<any, any>'.
+          Property '[Element]' is missing in type 'typeof MyComponent' but required in type 'AcceptsBlocks<any, any>'.",
+          "range": {
+            "end": {
+              "character": 43,
+              "line": 5,
+            },
+            "start": {
+              "character": 0,
+              "line": 5,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2345)",
+          "tags": [],
+        },
+        {
+          "message": "The {{component}} helper can't be used to directly invoke a component under Glint. Consider first binding the result to a variable, e.g. '{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as '<ComponentName @arg={{value}}>...</ComponentName>'.
+        Argument of type 'abstract new () => PartiallyAppliedComponent<{ message?: string | undefined; }, { message: string; }, AcceptsBlocks<FlattenBlockParams<{ default: { Params: { Positional: []; }; }; }>, null>>' is not assignable to parameter of type 'AcceptsBlocks<any, any>'.",
+          "range": {
+            "end": {
+              "character": 56,
+              "line": 6,
+            },
+            "start": {
+              "character": 0,
+              "line": 6,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(2345)",
+          "tags": [],
+        },
+      ]
+    `);
+  });
 });

--- a/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
+++ b/packages/core/__tests__/language-server/diagnostic-augmentation.test.ts
@@ -513,4 +513,126 @@ describe('Language Server: Diagnostic Augmentation', () => {
       ]
     `);
   });
+
+  test('unresolved globals', () => {
+    project.setGlintConfig({ environment: ['ember-loose'] });
+    project.write({
+      'index.ts': stripIndent`
+        import Component from '@glimmer/component';
+
+        export default class MyComponent extends Component {
+          declare locals: { message: string };
+        }
+      `,
+      'index.hbs': stripIndent`
+        {{! failed global lookups (custom message about the registry) }}
+        <Foo />
+        <foo.ok />
+        {{foo.bar}}
+        {{concat foo}}
+
+        {{#let this.locals as |locals|}}
+          {{! failed non-global lookup (no custom message) }}
+          {{locals.bad-thing}}
+        {{/let}}
+      `,
+    });
+
+    let server = project.startLanguageServer();
+    let diagnostics = server.getDiagnostics(project.fileURI('index.hbs'));
+
+    expect(diagnostics).toMatchInlineSnapshot(`
+      [
+        {
+          "message": "Unknown name 'Foo'. If this isn't a typo, you may be missing a registry entry for this value; see the Template Registry page in the Glint documentation for more details.
+        Element implicitly has an 'any' type because expression of type '\\"Foo\\"' can't be used to index type 'Globals'.
+          Property 'Foo' does not exist on type 'Globals'.",
+          "range": {
+            "end": {
+              "character": 7,
+              "line": 1,
+            },
+            "start": {
+              "character": 0,
+              "line": 1,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(7053)",
+          "tags": [],
+        },
+        {
+          "message": "Unknown name 'foo'. If this isn't a typo, you may be missing a registry entry for this value; see the Template Registry page in the Glint documentation for more details.
+        Element implicitly has an 'any' type because expression of type '\\"foo\\"' can't be used to index type 'Globals'.
+          Property 'foo' does not exist on type 'Globals'.",
+          "range": {
+            "end": {
+              "character": 10,
+              "line": 2,
+            },
+            "start": {
+              "character": 0,
+              "line": 2,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(7053)",
+          "tags": [],
+        },
+        {
+          "message": "Unknown name 'foo'. If this isn't a typo, you may be missing a registry entry for this value; see the Template Registry page in the Glint documentation for more details.
+        Element implicitly has an 'any' type because expression of type '\\"foo\\"' can't be used to index type 'Globals'.
+          Property 'foo' does not exist on type 'Globals'.",
+          "range": {
+            "end": {
+              "character": 9,
+              "line": 3,
+            },
+            "start": {
+              "character": 2,
+              "line": 3,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(7053)",
+          "tags": [],
+        },
+        {
+          "message": "Unknown name 'foo'. If this isn't a typo, you may be missing a registry entry for this value; see the Template Registry page in the Glint documentation for more details.
+        Element implicitly has an 'any' type because expression of type '\\"foo\\"' can't be used to index type 'Globals'.
+          Property 'foo' does not exist on type 'Globals'.",
+          "range": {
+            "end": {
+              "character": 12,
+              "line": 4,
+            },
+            "start": {
+              "character": 9,
+              "line": 4,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(7053)",
+          "tags": [],
+        },
+        {
+          "message": "Element implicitly has an 'any' type because expression of type '\\"bad-thing\\"' can't be used to index type '{ message: string; }'.
+        Property 'bad-thing' does not exist on type '{ message: string; }'.",
+          "range": {
+            "end": {
+              "character": 20,
+              "line": 8,
+            },
+            "start": {
+              "character": 4,
+              "line": 8,
+            },
+          },
+          "severity": 1,
+          "source": "glint:ts(7053)",
+          "tags": [],
+        },
+      ]
+    `);
+  });
 });

--- a/packages/transform/src/diagnostics/augmentation.ts
+++ b/packages/transform/src/diagnostics/augmentation.ts
@@ -110,6 +110,38 @@ function checkResolveError(
   diagnostic: Diagnostic,
   mapping: MappingTree
 ): ts.DiagnosticMessageChain | undefined {
+  // The diagnostic might fall on a lone identifier or a full path; if the former,
+  // we need to traverse up through the path to find the true parent.
+  let sourceMapping = mapping.sourceNode.type === 'Identifier' ? mapping.parent : mapping;
+  let parentNode = sourceMapping?.parent?.sourceNode;
+
+  // If this error is on the first param to a {{component}} invocation, this means
+  // we either have a non-component value or a string that's missing from the registry.
+  if (
+    (parentNode?.type === 'SubExpression' || parentNode?.type === 'MustacheStatement') &&
+    parentNode.path.type === 'PathExpression' &&
+    parentNode.path.original === 'component' &&
+    parentNode.params[0] === sourceMapping?.sourceNode
+  ) {
+    if (sourceMapping.sourceNode.type === 'StringLiteral') {
+      return addGlintDetails(
+        diagnostic,
+        `Unknown component name '${sourceMapping.sourceNode.value}'. If this isn't a typo, you may be ` +
+          `missing a registry entry for this name; see the Template Registry page in the Glint ` +
+          `documentation for more details.`
+      );
+    } else {
+      return addGlintDetails(
+        diagnostic,
+        `The type of this expression doesn't appear to be a valid value to pass the {{component}} ` +
+          `helper. If possible, you may need to give the expression a narrower type, ` +
+          `for example \`'component-a' | 'component-b'\` rather than \`string\`.`
+      );
+    }
+  }
+
+  // Otherwise if this is on a top level invocation, we're trying to use a template-unaware
+  // value in a template-specific way.
   let nodeType = mapping.sourceNode.type;
   if (nodeType === 'ElementNode' || nodeType === 'PathExpression' || nodeType === 'Identifier') {
     return addGlintDetails(

--- a/packages/transform/src/diagnostics/augmentation.ts
+++ b/packages/transform/src/diagnostics/augmentation.ts
@@ -58,7 +58,8 @@ function checkAssignabilityError(
     node.type === 'MustacheStatement' &&
     (parentNode.type === 'Template' ||
       parentNode.type === 'BlockStatement' ||
-      parentNode.type === 'ElementNode')
+      parentNode.type === 'ElementNode') &&
+    !(node.path.type === 'PathExpression' && node.path.original === 'yield')
   ) {
     // Otherwise, if it's on a full {{mustache}} and it's in a top-level position,
     // it's a DOM content type issue.

--- a/packages/transform/src/diagnostics/augmentation.ts
+++ b/packages/transform/src/diagnostics/augmentation.ts
@@ -56,14 +56,39 @@ function checkAssignabilityError(
         'If you want to set an event listener, consider using the `{{on}}` modifier instead.'
     );
   } else if (
+    node.type === 'BlockStatement' &&
+    node.path.type === 'PathExpression' &&
+    node.path.original === 'component'
+  ) {
+    // If it's attempted direct usage of `{{#component}}` as a curly block component,
+    // give a special note that that's not supported.
+    return addGlintDetails(
+      message,
+      `The {{component}} helper can't be used to directly invoke a component under Glint. ` +
+        `Consider first binding the result to a variable, e.g. ` +
+        `'{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as ` +
+        `'<ComponentName @arg={{value}}>...</ComponentName>'.`
+    );
+  } else if (
     node.type === 'MustacheStatement' &&
     (parentNode.type === 'Template' ||
       parentNode.type === 'BlockStatement' ||
       parentNode.type === 'ElementNode') &&
     !(node.path.type === 'PathExpression' && node.path.original === 'yield')
   ) {
-    // Otherwise, if it's on a full {{mustache}} and it's in a top-level position,
-    // it's a DOM content type issue.
+    // If we're looking at a top-level {{mustache}}, we first double check whether
+    // it's an attempted inline {{component 'foo'}} invocation...
+    if (node.path.type === 'PathExpression' && node.path.original === 'component') {
+      return addGlintDetails(
+        message,
+        `The {{component}} helper can't be used to directly invoke a component under Glint. ` +
+          `Consider first binding the result to a variable, e.g. ` +
+          `'{{#let (component 'component-name') as |ComponentName|}}' and then invoking it as ` +
+          `'<ComponentName @arg={{value}} />'.`
+      );
+    }
+
+    // Otherwise, it's a DOM content type issue.
     return addGlintDetails(
       message,
       'Only primitive values and certain DOM objects (see `ContentValue` in `@glint/template`) are ' +

--- a/packages/transform/src/diagnostics/augmentation.ts
+++ b/packages/transform/src/diagnostics/augmentation.ts
@@ -32,6 +32,7 @@ const diagnosticHandlers: Record<number, DiagnosticHandler | undefined> = {
   2554: adjustArgumentArity, // TS2554: Expected N arguments, but got M.
   2555: adjustArgumentArity, // TS2555: Expected at least N arguments, but got M.
   2769: checkResolveError, // TS2769: No overload matches this call.
+  7053: checkImplicitAnyError, // TS7053: Element implicitly has an 'any' type because expression of type '"X"' can't be used to index type 'Y'.
 };
 
 function checkAssignabilityError(
@@ -115,6 +116,39 @@ function checkResolveError(
       diagnostic,
       'The given value does not appear to be usable as a component, modifier or helper.'
     );
+  }
+}
+
+function checkImplicitAnyError(
+  diagnostic: Diagnostic,
+  mapping: MappingTree
+): ts.DiagnosticMessageChain | undefined {
+  let messageText =
+    typeof diagnostic.messageText === 'string'
+      ? diagnostic.messageText
+      : diagnostic.messageText.messageText;
+
+  // We don't want to bake in assumptions about the exact format of TS error messages,
+  // but we can assume that the name of the type we're indexing (`Globals`) will appear
+  // in the text in the case we're interested in.
+  if (messageText.includes('Globals')) {
+    let { sourceNode } = mapping;
+
+    // This error may appear either on `<Foo />` or `{{foo}}`/`(foo)`
+    let globalName =
+      sourceNode.type === 'ElementNode'
+        ? sourceNode.tag.split('.')[0]
+        : sourceNode.type === 'PathExpression' && sourceNode.head.type === 'VarHead'
+        ? sourceNode.head.name
+        : null;
+
+    if (globalName) {
+      return addGlintDetails(
+        diagnostic,
+        `Unknown name '${globalName}'. If this isn't a typo, you may be missing a registry entry ` +
+          `for this value; see the Template Registry page in the Glint documentation for more details.`
+      );
+    }
   }
 }
 

